### PR TITLE
[ARM32/Linux] Initial bring up of FEATURE_INTERPRETER

### PR DESCRIPTION
**Pull request by mskvortsov**
During the investigation of a performance implications of the CoreCLR's interpreter I made some fixes to bring it up. Issue #8496 (RyuJIT ARM32 Backend) can possibly benefit from this PR too: turning on the interpreter on a not-yet fully codegen-ed methods can simplify testing a bit.

After the patch is applied, adding something like
```
#if defined(_TARGET_ARM_)
#define FEATURE_INTERPRETER
#endif
```
at the end of `src/vm/switches.h` will enable the interpreter.

I've tested it on a simple HelloWorld and recently on the ByteMark test. Some methods should be excluded with COMPlus_InterpretExclude option though, since further work is needed. I am aware that some of the fixes proposed are dirty hacks actually. I would appreciate any critics and suggestions for a better solution.

BTW, interp's perf is suprisingly great, see the ByteMark stats here: [arm32-bytemark-interp.txt](https://gist.github.com/mskvortsov/fd61baa179bb510ab0d6cc850e43c93a).

### CODE REVIEW

**Comment by mskvortsov** [comment]
> BTW, interp's perf is suprisingly great

Ok, `COMPlus_InterpreterDoLoopMethods=1` and `COMPlus_InterpreterJITThreshold=-1` did the trick: I finally got not at all so surprising huge slow down factor of several hundred of times.

File: src/vm/stublink.cpp [hunk 1]
```diff
@@ -2201,6 +2205,14 @@ Stub* Stub::NewStub(PTR_VOID pCode, DWORD flags)
         pBlock = new (executable) BYTE[totalSize];
 #else
         pBlock = new BYTE[totalSize];
+#if defined(PLATFORM_UNIX) && defined(_ARM_)
+        _ASSERTE(totalSize <= 0x1000);
```
Comment (jkotas) [inline src/vm/stublink.cpp]: This hack is really problematic from security point of view. It will flip random parts of the regular head to executable.

**Comment by mskvortsov** [comment]
@Dmitri-Botcharnikov

File: src/vm/stublink.cpp (hunk 1 above)
Comment (mskvortsov) [inline src/vm/stublink.cpp]: Thanks. Now I've found a proper way to allocate stubs on the executable heap. Please take a look at the updated commit.

File: src/vm/interpreter.cpp [hunk 2]
```diff
@@ -10316,15 +10321,19 @@ void Interpreter::CallI()
             }
             else
             {
-                pMD = g_pPrepareConstrainedRegionsMethod;  // A random static method.
+                pMD = g_pExecuteBackoutCodeHelperMethod;  // A random static method.
             }
             MethodDescCallSite mdcs(pMD, &mSig, ftnPtr);
+#if 0
```
Comment (jkotas) [inline src/vm/interpreter.cpp]: Why is this `#if 0` out?

**Comment by jkotas** [comment]
LGTM otherwise.

File: src/vm/interpreter.cpp (hunk 2 above)
Comment (mskvortsov) [inline src/vm/interpreter.cpp]: It triggers assertion in the Thread::ObjectRefFlush function at [threads.cpp:6861](https://github.com/dotnet/coreclr/blob/master/src/vm/threads.cpp#L6861):
```c++
 _ASSERTE(thread->PreemptiveGCDisabled());  // Should have been in managed code
```
I don't know much about GC subsystem and preemption in particular, so I just disabled that code.

**Comment by jkotas** [comment]
> BTW, interp's perf is suprisingly great, see the ByteMark stats here

Are you sure that it is actually running through the interpreter?

**Comment by mskvortsov** [comment]
I'm sure it's not. I've wrote about this earlier already: I missed the `COMPlus_InterpreterDoLoopMethods=1` option for that run. Slowdown is huge, I just needed an evidence for that.

(Initially I was investigating some ways to reduce CoreCLR's memory consumption and interp was obvious candidate to try out. Later I measured a typical ratio of emitted code size to the total resident memory size, and it turned out to be less than 5%. And less than 2% for relatively lightweight interactive Xamarin apps).

File: src/vm/interpreter.cpp (hunk 2 above)
Comment (mskvortsov) [inline src/vm/interpreter.cpp]: Ok, done.

**Comment by jkotas** [comment]
Thanks!

**Pull request merged**